### ISSUE TITLE
getPageRegion top is not returning correct value

### ISSUE THREAD

**Issue by jezerinac** [DDMAL/diva.js#371]
![image](https://cloud.githubusercontent.com/assets/6097594/19336632/eaad1806-90c1-11e6-86aa-1e091cc8a7b4.png)

Here's my test code:

diva.Events.subscribe('ViewerDidScroll', function() {
            var pageOffset0 = dv.getPageOffset(0, {incorporateViewport:true});
            window.console.log('pageOffset:0');
            window.console.log(pageOffset0);
            var pageOffset1 = dv.getPageOffset(1, {incorporateViewport:true});
            window.console.log('pageOffset:1');
            window.console.log(pageOffset1);
        }, dv.getSettings().ID);

**Comment by wabain** [comment]
This is actually working as intended. `getPageOffset` gives coordinates relative to the entirety of diva's scrollable area, not relative to the viewport. The `incorporateViewport` option is not actually as helpful as it sounds. It just handles a special case where the extent of the viewport is greater than the logical size of diva's document. If you're overlaying something on top of the viewer you need this to get the right coordinates to position something within the diva-inner element, but the internals for which the underlying `getPageRegion` method were implemented don't need to care about this because they just work with the logical document. It's a fairly low-level implementation detail that should have been handled more gracefully.

The variables which actually let you work with viewport-relative coordinates are `settings.horizontalOffset` and `settings.verticalOffset`, but they use a confusing coordinate system. They give the distance from the top left of the current page to the center of the viewport. There's also `settings.viewport.top` and `settings.viewport.left`, which you could use along with `getPageOffset` to get a viewport relative offset. Those aren't documented but if you open an issue for it it should be fine.

**Comment by jezerinac** [comment]
What I'm trying to do is the basically get the rendering position of a page x,y. So given a page position say (301,720) to get the rendered position (x,y).

Example:
`xy = getRenderedXy(pageIndex, pageX, pageY);`

I can do this using internal code, basically the reverse of my open pull request #369.
I was hoping I do the same calculation with existing public API methods.
@wabain do you think this is possible with existing public methods?  Thanks very much for your help.

Here's sample code for the calculation by percent:
`
Renderer.prototype.getRenderedXyByPercentage = function (pageIndex, percentageX, percentageY)
{
    for (var i = 0; i < this._renderedPages.length; i++)
    {
        // this check ensures the page is actually visible
        if (this._renderedPages[i] != pageIndex)
        {
            continue;
        }

        var pageInfo = this.layout.getPageInfo(pageIndex);
        var pageOffset = this._getImageOffset(pageIndex);

        var viewportPaddingX = Math.max(0, (this._viewport.width - this.layout.dimensions.width) / 2);
        var viewportPaddingY = Math.max(0, (this._viewport.height - this.layout.dimensions.height) / 2);

        var viewportOffsetX = pageOffset.left - this._viewport.left + viewportPaddingX;
        var viewportOffsetY = pageOffset.top - this._viewport.top + viewportPaddingY;

        var destXOffset = viewportOffsetX < 0 ? -viewportOffsetX : 0;
        var destYOffset = viewportOffsetY < 0 ? -viewportOffsetY : 0;

        var canvasX = Math.max(0, viewportOffsetX);
        var canvasY = Math.max(0, viewportOffsetY);

        var x = (percentageX * pageInfo.dimensions.width) + canvasX - destXOffset;
        var y = (percentageY * pageInfo.dimensions.height) + canvasY - destYOffset;

        return {x:x, y:y};
    }
    return null;
};
`

**Comment by wabain** [comment]
Untested, but roughly equivalent:

```js
function getViewportOffsetByPercentagePageOffset(dv, pageIndex, xPercent, yPercent)
{
    // Exit early if the page isn't visible at all. Note that this isn't sufficient
    // to make sure that the point that's returned is actually within the viewport.
    if (!dv.isPageInViewport(pageIndex))
        return null;

    var viewport = dv.getSettings().viewport;
    var pageDims = dv.getPageDimensions(pageIndex);

    // Get the offset of the page relative to Diva's scrollable area, including the
    // extra viewport margins if the viewport extends further than the logical
    // document, and excluding padding around the page so that only the page
    // image is included
    var absolutePageOffset = dv.getPageOffset(pageIndex, {
        excludePadding: true,
        incorporateViewport: true
    });

    return {
        x: (absolutePageOffset.left - viewport.left) + (pageDims.width * xPercent),
        y: (absolutePageOffset.top - viewport.top) + (pageDims.height * yPercent)
    };
}
```

**Comment by jezerinac** [comment]
works, thanks!

**Comment by jezerinac** [comment]
Sorry, the code given does NOT work when in the grid view.

**Comment by wabain** [comment]
What outcome do you see in grid view?

**Comment by jezerinac** [comment]
![image](https://cloud.githubusercontent.com/assets/6097594/19704906/9f9fc8de-9abe-11e6-8e95-4dc137fe01b1.png)

![image](https://cloud.githubusercontent.com/assets/6097594/19704958/dedb7156-9abe-11e6-9d5e-de85983f20d5.png)

**Comment by wabain** [comment]
From a quick look, I think this is a bug which occurs when the extent of the viewport is greater than the extent of the document along the primary axis. Do things work as expected if you view the grid with 4 pages per row?

**Comment by jezerinac** [comment]
4 rows per page still has the problem but if the pages per row is changed to remove all vertical padding it aligns up properly.

**Comment by EricHanLiu** [comment]
getPageRegion was recently slightly changed to fix issue #436 so the top and bottom returns are more correct now.

Closing, inactive
